### ISSUE TITLE
Script generator location

### ISSUE THREAD

**Issue by outpaddling** [PacificBiosciences/FALCON#177]
Where are scripts such as 0-rawreads/prepare_rdb.sh generated?

I'm trying to run with job_type = SLURM and SLURM is rejecting the scripts because they are missing a shebang line.

If you can point me to the right place, I'll patch it to include a portable shebang line.

Thanks,

    Jason

**Comment by outpaddling** [comment]
Never mind, I found it by grepping all the python scripts for "export".  I have a multinode job running now, although we haven't figured out how to control the thread count per job yet.  I assume the SGE smp flag has something to do with this.  I don't think there's a similar flag in SLURM.

**Comment by outpaddling** [comment]
A more specific question: Is there a way to control the number of threads per job when using job_type = local?

If so, can you update the Falcon manual to include it?

Thanks.

**Comment by pb-cdunn** [comment]
I don't think we have any shebang lines. What's wrong with `bash script.sh`?

I don't think we have any limit on local threads in pypeFLOW. If you have a problem locally, you probably need distributed jobs. The sub-scripts, however, provide `--n_core=N` to limit the number of threads.

**Comment by pb-jchin** [comment]
The daligner uses 4 thread and it is hard-coded.  Pypeflow monitors each indpendent job per thread. You can control that by changing `*concurrent_jobs`. The number of threads of the other independent task can be controlled by `n_core` as described by @pb-cdunn

**Comment by outpaddling** [comment]
On 08/28/15 05:04, Jason Chin wrote:
>
> The daligner uses 4 thread and it is hard-coded. Pypeflow monitors 
> each indpendent job per thread. You can control that by changing 
> |*concurrent_jobs|. The number of threads of the other independent 
> task can be controlled by |n_core| as described by @pb-cdunn 
> <https://github.com/pb-cdunn>
>
Thanks, Jason.  Is daligner hard-coded at 4 because that was deemed 
optimal, or would it benefit from more flexibility?  We'll be running on 
a large shared HPC system with 16-24 cores per node, so in theory we 
could run anything from sixteen to twenty-four 1-core jobs to one 16- to 
24-core job.  I would consider patching the source myself if different 
multithreading levels might benefit performance.

**Comment by pb-cdunn** [comment]
It's 4 because those 4 jobs share memory for the input. Extra threads could hurt because they could use up precious processor cache-space.

**Comment by outpaddling** [comment]
Nothing wrong with "bash script" in general, but that's not how SLURM invokes them behind the scenes.  It's shell-agnostic and relies on shebang lines.  I already created a patch for this and got it running.

We'll experiment with --n_core, thanks.

FYI, I ran on a 12-core node with both 1 and 3 daligner jobs and saw nearly a speedup of about 2.2 when running 3 jobs.  I wouldn't expect the results to be much different with a single 12-core job.  I might hack the code to use 16 (nthreads has to be a power of 2 the way daligner is written) and try this out next week on a 16-core node.

Regards,

    Jason
